Fix TypeScript

catdevnull opened the pull request:
Use tsimp instead of ts-node and make sure all imports follow actual ESM

Similar to https://github.com/hyphacoop/social.distributed.press/pull/65

catdevnull committed:
Fix TypeScript

Use tsimp instead of ts-node and make sure all imports follow actual ESM

catdevnull commented:
Marked as draft as I can't get that IPFS type error fixed. I believe it's a bug in the upstream libraries.

RangerMauve commented:
Maybe you could override the types somehow? Does the code run if you ignore type errors?

catdevnull commented:
>Maybe you could override the types somehow? Does the code run if you ignore type errors?

Type errors can't be ignored with tsimp, as it uses the official TypeScript compiler (tsc). That's why I suggested using something like [tsx](https://github.com/privatenumber/tsx), which uses `esbuild`, which just transpiles TypeScript into JavaScript without typechecking. Then, typechecking can be done by a CI or a watch script, along with tests/linting.

I'll try to override the types.

catdevnull commented:
I'm trying to work on this but dist.ipfs.tech is down since yesterday, so `npm install` on `go-ipfs` doesn't work :|

catdevnull committed:
fix: stop fetching id when starting the API

this causes types error and also is already done by every controller in ipfsd-ctl

catdevnull committed:
fix: expect error in addAll()

sadly it seems the types in this package are bad, and we're using an outdated version. I tried upgrading the packge but it seems that there's a ton of API changes, so I'll leave it like this for now.

catdevnull commented:
I think these changes are OK for now. Either way we should upgrade the `ipfsd-ctl` package at some point.

catdevnull commented:
Just noticed that while the TS is fine, the linting on the IPFS file is broken. As I mentioned we should probably migrate off of [ipfsd-ctl](https://github.com/ipfs/js-ipfsd-ctl) as its types are broken and the readme says
>It is designed mostly for testing interoperability and is **not suitable for production use.**
